python/taichi/examples/simulation/laplace_equation.py

```diff
@@ -0,0 +1,307 @@
+                if boundary == 0:
+                    points[index] = vec2(-dt * refillVelThresh,
+                                         (i + .5) / screen[1])
+                elif boundary == 1:
+                    points[index] = vec2((i + .5) / screen[0],
+                                         -dt * refillVelThresh)
+                elif boundary == 2:
+                    points[index] = vec2(1 + dt * refillVelThresh,
+                                         (i + .5) / screen[1])
+                elif boundary == 3:
+                    points[index] = vec2((i + .5) / screen[0],
+                                         1 + dt * refillVelThresh)
+                break
+            index += 1
+            if index >= maxPoints:
+                index = 0
+        if not found:
+            break
+    return index
+
+
+@ti.kernel
+def refillPoints():
+    # traverse positions and refill points on the boundary.
+    # if the normal velocity is less than thresh, refill point will be deleted when update.
+    index = 0
+    found = True
+    ti.loop_config(serialize=True)
+    for i in range(4):
+        _index = refillPointsOnOneBoundary(i, index)
+        if _index == index:
+            break
+        else:
+            index = _index
+
+
+@ti.kernel
+def updatePoints():
+    for i in points:
+        if points[i][0] != -100:
+            vel = getVel(points[i])
+            points[i] += vel * dt
+        if not (0 <= points[i][0] <= 1 and 0 <= points[i][1] <= 1):
+            points[i] = vec2(-100, -100)
+        for j in range(maxElements):
+            if sources[j].q < 0 and ti.math.length(points[i] -
+                                                   sources[j].pos) < 0.025:
+                points[i] = vec2(-100, -100)
+            if dipoles[j].m != 0 and ti.math.length(points[i] -
+                                                    dipoles[j].pos) < 0.05:
+                points[i] = vec2(-100, -100)
+
+
+@ti.kernel
+def updateArrows():
+    for x, y in ti.ndrange(*arrowField):
+        vel = getVel(arrows[x, y].base)
+        arrows[x, y].vel = vel.norm()
+        arrows[x, y].dir = vel / vel.norm() / meshSpace / 1.5
+
+
+def drawArrows(gui):
+    global fade
+    if fade < fadeMax:
+        fade += 1
+    if fade == 0:
+        updateArrows()  # after fade out, update arrow filed
+    else:
+        arr = arrows.to_numpy()
+        vel = arr['vel'].reshape(1, -1)[0]
+        vel = (vel / vel.max() * 0xdd + 0x11) * (math.fabs(fade / fadeMax))
+        mean = vel.mean()
+        if mean > 0x7f:
+            vel /= mean / 0x7f  # make uniform stream more beautiful
+        vel = vel.astype(int)
+        vel *= 2**16 + 2**8 + 1
+        gui.arrows(
+            arr['base'].reshape(arr['base'].shape[0] * arr['base'].shape[1],
+                                2),
+            arr['dir'].reshape(arr['dir'].shape[0] * arr['dir'].shape[1], 2),
+            radius=1.5,
+            color=vel)
+
+
+def drawMark(gui, frame):
```

Comment:
Could you please help make the `gui` and `frame` global to make the pre-commit happy? ( though the error it raised is indeed weird.. lol

Reply:
ok. I replace them with `gui_` and `frame_count`. Hope it work.